### PR TITLE
feat/pro-609/order-validity-range

### PR DESCRIPTION
**Pull request by Janislav**
# Pull Request

Closes: PRO-615

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have updated documentation where appropriate.

## Summary

*Please include a succinct description of the purpose and content of the PR. What problem does it solve, and how? Link issues, discussions, other PRs, and anything else that will help the reviewer.*

### CODE REVIEW

File: state-chain/pallets/cf-pools/src/lib.rs [hunk 1]
```diff
@@ -472,6 +532,26 @@ pub mod pallet {
 					}
 				}
 			}
+			let scheduled_limit_orders = ScheduledLimitOrders::<T>::take(current_block);
+			for order_update in scheduled_limit_orders {
```
Comment (dandanlen) [inline state-chain/pallets/cf-pools/src/lib.rs:536]: Use `for OrderUpdate { lp, id, call } ` and you can avoid some unnecessary clones.

File: state-chain/pallets/cf-pools/src/lib.rs [hunk 2]
```diff
@@ -472,6 +532,26 @@ pub mod pallet {
...
+				if let Err(err) = order_update
+					.clone()
+					.call
```
Comment (dandanlen) [inline state-chain/pallets/cf-pools/src/lib.rs:539]: I think we might want to wrap the dispatch in a storage transaction to ensure we roll back state if it fails.

File: state-chain/pallets/cf-pools/src/lib.rs [hunk 3]
```diff
@@ -217,11 +218,65 @@ pub mod pallet {
...
+		pub option_tick: Option<Tick>,
+		pub sell_amount: AssetAmount,
+	}
+
+	/// Represents the operational details of a scheduled limit order.
+	#[derive(Clone, Debug, TypeInfo, PartialEq, Eq, Encode, Decode, MaxEncodedLen)]
+	#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+	pub struct OrderScheduleDetails<BlockNumber: PartialOrd + Copy> {
+		/// The block number range during which the order is valid on entering state-chain.
+		pub valid_at: Option<Range<BlockNumber>>,
```
[the remark this answers is not in the document]
Comment (Janislav) [reply state-chain/pallets/cf-pools/src/lib.rs:249]: Okay, I was thinking this is for checking when the transaction is actual entering the state-chain. When we craft the transaction we don't know when it finally gets picked up from the mem pool so we need some kind of validity checking if the transaction is still in the validity range. I can remeber we talked about that we want to keep it for that use case 🤔

File: api/bin/chainflip-lp-api/src/main.rs [hunk 4]
```diff
@@ -140,6 +141,7 @@ pub trait Rpc {
 		id: OrderIdJson,
 		tick: Option<Tick>,
 		amount_change: IncreaseOrDecrease<NumberOrHex>,
+		details: Option<OrderScheduleDetails<BlockNumber>>,
```
Comment (AlastairHolmes) [inline api/bin/chainflip-lp-api/src/main.rs:144]: Shouldn't this be called something like scheduling, or schedule_details..., details is abit unclear, given these are visible to external users via the jsonrpcs.

File: api/lib/src/lp.rs [hunk 5]
```diff
@@ -241,20 +243,39 @@ pub trait LpApi: SignedExtrinsicApi {
...
-				option_tick,
-				amount_change,
-			})
-			.await
-			.until_in_block()
-			.await?;
-
+		let events = if let Some(details) = details {
+			let (_tx_hash, events, ..) = self
+				.submit_signed_extrinsic(pallet_cf_pools::Call::schedule {
```
Comment (AlastairHolmes) [inline api/lib/src/lp.rs:251]: You would dedup some codeby moving the update_limit_order construction outside the `if let Some()`

Comment (AlastairHolmes) [reply api/lib/src/lp.rs:251]: You could then dedup more code, by introducing a function scheduled_or_immediate (or something like that) that takes a call, and an Option<OrderScheduleDetail>, and either schedules it or does the call immediately.

File: state-chain/pallets/cf-pools/src/lib.rs [hunk 6]
```diff
@@ -472,6 +532,26 @@ pub mod pallet {
...
+						order_id: order_update.id,
+						error: err.error,
+					});
+				} else {
+					Self::deposit_event(Event::<T>::SuccessfullyExecutedLimitOrder {
+						lp: order_update.lp,
+						order_id: order_update.id,
+					});
+				}
+			}
```
[the remark this answers is not in the document]
Comment (Janislav) [reply state-chain/pallets/cf-pools/src/lib.rs:554]: Or we remove all the manual calculation and just create a benchmark for it.

File: api/bin/chainflip-lp-api/src/main.rs (hunk 4 above)
Comment (Janislav) [reply api/bin/chainflip-lp-api/src/main.rs:144]: Good point 👍

File: state-chain/pallets/cf-pools/src/lib.rs [hunk 7]
```diff
@@ -1116,31 +1071,25 @@ pub mod pallet {
...
 		/// - [BadOrigin](frame_system::BadOrigin)
 		/// - [UnsupportedCall](pallet_cf_pools::Error::UnsupportedCall)
-		/// - [OrderScheduleDetailsInconsistent](pallet_cf_pools::Error::OrderScheduleDetailsInconsistent)
 		#[pallet::call_index(8)]
 		#[pallet::weight(T::WeightInfo::schedule())]
 		pub fn schedule(
 			origin: OriginFor<T>,
 			call: Box<Call<T>>,
-			details: OrderScheduleDetails<BlockNumberFor<T>>,
+			dispatch_at: BlockNumberFor<T>,
```
Comment (dandanlen) [inline state-chain/pallets/cf-pools/src/lib.rs:1082]: We don't handle expired blocks (in the past) or if it's the current block.

**Review by dandanlen — COMMENTED** [review]
I added checks to handle dispatch_at when it's in the past (error) or when it's the current block (dispatch immediately). 

I also did some renaming to avoid suggesting order executions. 


Btw. this will need a companion PR in the docs repo.

File: state-chain/pallets/cf-pools/src/lib.rs (hunk 7 above)
Comment (dandanlen) [reply state-chain/pallets/cf-pools/src/lib.rs:1082]: (fixed)

**Comment by martin-chainflip** [comment]
This looks ok, but it only applies to limit orders and only supports a starting time. In the corresponding ticket PRO 615 we were discussing any LP order (including range orders) and also an endpoint on which the order expires automatically. These features are not in the PR?

**Review by martin-chainflip — APPROVED** [review]

**Comment by Janislav** [comment]
Added docs PR here: https://github.com/chainflip-io/chainflip-docs/pull/114

**Comment by martin-chainflip** [comment]
LGTM?

**Comment by dandanlen** [comment]
Yep let's get this merged.

**Pull request merged**